### PR TITLE
SYN-6134: storm http proxy setting permission gated

### PR DESCRIPTION
**Pull request by MichaelSquires**
feat: Added `storm.lib.inet.http.proxy` permission to `$lib.inet.http` APIs so non-admin users can be granted access to specify a proxy server.
feat: Added `storm.lib.inet.http.proxy` permission to `$lib.axon.wget` and `$lib.axon.wput` APIs to non-admin users can be granted access to specify a proxy server.

### CODE REVIEW

File: synapse/lib/stormtypes.py [hunk 1]
```diff
@@ -2128,6 +2127,9 @@ async def wput(self, sha256, url, headers=None, params=None, method='PUT', ssl=T
         params = self.strify(params)
         headers = self.strify(headers)
 
+        if proxy is not None and not self.runt.allowed(('axon', 'proxy')):
+            raise s_exc.AuthDeny(mesg=s_exc.proxy_admin_mesg, user=self.runt.user.iden, username=self.runt.user.name)
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2131]: the proxy_admin_mesg value is no longer correct.

Can we use `confirm()` to generate the correct permissions error?

File: synapse/lib/stormtypes.py [hunk 2]
```diff
@@ -2128,6 +2127,9 @@ async def wput(self, sha256, url, headers=None, params=None, method='PUT', ssl=T
         params = self.strify(params)
         headers = self.strify(headers)
 
+        if proxy is not None:
+            self.runt.confirm(('axon', 'proxy'))
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2131]: Is this a permission we need to be checking on the AxonApi too?

File: synapse/tests/test_lib_stormhttp.py [hunk 3]
```diff
@@ -544,6 +546,44 @@ async def test_storm_http_proxy(self):
...
+            await visi.addRule((True, ('axon', 'proxy')))
+
+            resp = await core.callStorm(q1, opts=asvisi)
+            self.eq('ProxyConnectionError', resp['err'][0])
+
+            resp = await core.callStorm(q2, opts=asvisi)
+            self.eq('ProxyConnectionError', resp['err'][0])
+
+            resp = await core.callStorm(q3, opts=asvisi)
+            self.eq('ProxyConnectionError', resp['err'][0])
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_stormhttp.py:585]: Can we add a test for `proxy=$lib.false` that allows the low-privileged user to bypass the Cortex `http:proxy` configuration when they have the correct permission?

File: synapse/lib/stormhttp.py [hunk 4]
```diff
@@ -250,6 +250,10 @@ class LibHttp(s_stormtypes.Lib):
                   'returns': {'type': 'str', 'desc': 'The reason phrase for the status code.', }}},
     )
     _storm_lib_path = ('inet', 'http')
+    _storm_lib_perms = (
+        {'perm': ('inet', 'http', 'proxy'), 'gate': 'cortex',
+         'desc': 'Permits a user to use a proxy with `$lib.inet.http` APIs.'},
```
Comment (vEpiphyte) [inline synapse/lib/stormhttp.py:255]: Maybe something like ``Permits a user to specify the proxy used with `$lib.inet.http` APIs`` ?

File: synapse/lib/stormtypes.py (hunk 2 above)
Comment (MichaelSquires) [reply synapse/lib/stormtypes.py:2131]: Yes, good catch, thanks! :)

File: synapse/tests/test_lib_stormhttp.py [hunk 5]
```diff
@@ -544,6 +546,64 @@ async def test_storm_http_proxy(self):
             self.false(resp.get('ok'))
             self.isin('connect to proxy 127.0.0.1:1', resp['mesg'])
 
+        async with self.getTestCore() as core:
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_stormhttp.py:549]: Should this have a `conf = {'http:proxy': 'socks5://user:pass@127.0.0.1:1'}` value?

File: synapse/lib/stormhttp.py [hunk 6]
```diff
@@ -250,6 +250,10 @@ class LibHttp(s_stormtypes.Lib):
                   'returns': {'type': 'str', 'desc': 'The reason phrase for the status code.', }}},
     )
     _storm_lib_path = ('inet', 'http')
+    _storm_lib_perms = (
+        {'perm': ('inet', 'http', 'proxy'), 'gate': 'cortex',
```
Comment (vEpiphyte) [inline synapse/lib/stormhttp.py:254]: should discuss if this should be `storm.lib.inet.http.proxy` to be in line with #3135  changes

File: synapse/lib/stormtypes.py [hunk 7]
```diff
@@ -2091,6 +2090,9 @@ async def wget(self, url, headers=None, params=None, method='GET', json=None, bo
         timeout = await toprim(timeout)
         proxy = await toprim(proxy)
 
+        if proxy is not None:
+            self.runt.confirm(('storm', 'lib', 'axon', 'proxy'))
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2094]: I thought we were moving this to `storm.lib.inet.http.proxy` as a single permission for controlling proxy bypass?

File: synapse/lib/stormtypes.py [hunk 8]
```diff
@@ -2112,9 +2114,6 @@ async def wput(self, sha256, url, headers=None, params=None, method='PUT', ssl=T
         if not self.runt.allowed(('axon', 'wput')):
             self.runt.confirm(('storm', 'lib', 'axon', 'wput'))
 
-        if proxy is not None and not self.runt.isAdmin():
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2115]: I thought we were moving this to storm.lib.inet.http.proxy as a single permission for controlling proxy bypass?

File: synapse/lib/stormtypes.py [hunk 9]
```diff
@@ -2128,6 +2127,9 @@ async def wput(self, sha256, url, headers=None, params=None, method='PUT', ssl=T
         params = self.strify(params)
         headers = self.strify(headers)
 
+        if proxy is not None:
+            self.runt.confirm(('storm', 'lib', 'axon', 'proxy'))
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2131]: I thought we were moving this to storm.lib.inet.http.proxy as a single permission for controlling proxy bypass?

File: synapse/lib/stormtypes.py [hunk 10]
```diff
@@ -2001,6 +2001,8 @@ class LibAxon(Lib):
             'desc': 'Controls the ability to retrieve a file from URL and store it in the Axon.'},
         {'perm': ('storm', 'lib', 'axon', 'wput'), 'gate': 'cortex',
             'desc': 'Controls the ability to push a file from the Axon to a URL.'},
+        {'perm': ('storm', 'lib', 'axon', 'proxy'), 'gate': 'cortex',
+            'desc': 'Permits a user to specify the proxy used with `$lib.axon.wget` and `$lib.axon.wput` APIs.'},
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2005]: This would be OBE if we moved this to a single permission

File: synapse/lib/stormtypes.py (hunk 7 above)
Comment (MichaelSquires) [reply synapse/lib/stormtypes.py:2094]: I must've missed that part of the convo.

**Review by vEpiphyte — APPROVED** [review]

**Pull request merged**